contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -11,66 +11,150 @@ contract KeystoneForwarder is IForwarder, ConfirmedOwner, TypeAndVersionInterfac
   error ReentrantCall();
 
   /// @notice This error is returned when the data with report is invalid.
-  /// This can happen if the data is shorter than SELECTOR_LENGTH + REPORT_LENGTH.
+  /// This can happen if the data is shorter than SELECTOR_LENGTH + REPORT_HEADER_LENGTH.
   /// @param data the data that was received
   error InvalidData(bytes data);
 
+  /// @notice This error is thrown whenever trying to set a config
+  /// with a fault tolerance of 0
+  error FaultToleranceMustBePositive();
+
+  /// @notice This error is thrown whenever a report is signed
+  /// with more than the max number of signers
+  /// @param numSigners The number of signers who have signed the report
+  /// @param maxSigners The maximum number of signers that can sign a report
+  error ExcessSigners(uint256 numSigners, uint256 maxSigners);
+
+  /// @notice This error is thrown whenever a report is signed
+  /// with less than the minimum number of signers
+  /// @param numSigners The number of signers who have signed the report
+  /// @param minSigners The minimum number of signers that need to sign a report
+  error InsufficientSigners(uint256 numSigners, uint256 minSigners);
+
+  error RepeatedSigner();
+
+  error WrongNumberOfSignatures();
+
+  error InvalidDonId();
+  error InvalidSigner();
+
   uint256 private constant SELECTOR_LENGTH = 4;
-  uint256 private constant REPORT_LENGTH = 64;
 
-  struct HotVars {
-    bool reentrancyGuard; // guard against reentrancy
-  }
+  bool internal s_reentrancyGuard; // guard against reentrancy
 
-  HotVars internal s_hotVars; // Mixture of config and state, commonly accessed
+  /// @notice Contains the signing address of each oracle
+  struct OracleSet {
+    uint8 f; // Number of faulty nodes allowed
+    address[] signers;
+    mapping(address => uint256) _positions; // 1-indexed to detect unset values
+  }
+  mapping(bytes4 donId => OracleSet) internal s_configs;
 
-  mapping(bytes32 => address) internal s_reports;
+  // reportId = keccak256(bytes20(receiver) | selector | workflowExecutionId)
+  mapping(bytes32 reportId => address transmitter) internal s_reports;
 
   constructor() ConfirmedOwner(msg.sender) {}
 
-  // send a report to targetAddress
+  uint256 internal constant MAX_ORACLES = 31;
+
+  // NOTE: we don't inherit OCR2Base since unlike aggregator we only care about signers, not transmitters
+  // and the signers don't fetch their config from the forwarder
+  function setConfig(bytes4 donId, uint8 f, address[] calldata signers) external nonReentrant {
+    if (f == 0) revert FaultToleranceMustBePositive();
+    if (signers.length > MAX_ORACLES) revert ExcessSigners(signers.length, MAX_ORACLES);
+    if (signers.length <= 3 * f) revert InsufficientSigners(signers.length, 3 * f + 1);
+
+    // TODO: how does setConfig handle expiration? e.g. if the signer set changes
```

Comment:
Nit: I typically expect to see a link to the Jira ticket if you leave the TODO in the code. Otherwise, these aren't visible in Jira.

Reply:
@archseer what do you mean by expiration?

Reply:
In OCR2Aggregator changing the config also expires any past in-flight transmissions since the config digest will change (and needs to match on transmit).

Reply:
We're not sending the digest alongside our reports are we? I thought we will only rely purely on signature validation.

Reply:
We don't but the question is do we want something equivalent for expiration?

Reply:
Created a ticket to investigate: https://smartcontract-it.atlassian.net/browse/KS-202?atlOrigin=eyJpIjoiMGM1NGQ4ZTFmMDg1NDZlZjgxOWIyNDA4MjUzZGJjOTYiLCJwIjoiaiJ9